Handle location services disabled when requesting ENs

devinjameson opened the pull request:
Why: some Android devices require location services to be enabled for exposure notifications to work. If someone is on such a device and tries to enable exposure notifications with location services disabled, would like to show an alert that tells them to enable location services.

This commit:
- Adds `LocationOffAndRequired` to the list of possible states for the `ENPermissionStatus`
- If `ENPermissionStatus` is `LocationOffAndRequired`, show an alert to the user telling them to enable location services

![Screen Shot 2020-12-02 at 2 22 53 PM](https://user-images.githubusercontent.com/39350030/100920979-e883c800-34a9-11eb-9d02-e9abb58701f3.png)

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson commented on src/locales/en.json:
```diff
@@ -257,8 +257,8 @@
-    "location_alert_body": "To use Bluetooth scanning, the device location setting needs to be turned on.",
-    "location_alert_header": "Enable Location access from \"{{applicationName}}\"",
+    "location_alert_body": "To use Exposure Notifications, Location Services must be enabled",
+    "location_alert_header": "Enable Location Services",
```
@juancruzgs Can you confirm this copy is correct?

juancruzgs replied:
Looks good to me 👍

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson commented on src/useRequestExposureNotifications.tsx:
```diff
@@ -13,46 +13,34 @@ export const useRequestExposureNotifications = (): (() => void) => {
-    if (Platform.OS === "ios") {
-      showShareExposureInformationAlert()
+    if (response.kind === "success") {
+      switch (response.status) {
```
@mattThousand I'm confused as to why `BluetoothOff` is under the "success" column here, but I figured `LocationOff` should go here as well. Should this be refactored so that anything but `Active` is considered a "failure"?

These types seem weird to me considering the overlap:

```
type RequestAuthorizationError =
  | "IOSUnknown"
  | "BadParameter"
  | "NotEntitled"
  | "NotAuthorized"
  | "Unsupported"
  | "Invalidated"
  | "BluetoothOff"
  | "InsufficientStorage"
  | "NotEnabled"
  | "APIMisuse"
  | "Internal"
  | "InsufficientMemory"
  | "RateLimited"
  | "Restricted"
  | "BadFormat"
  | "DataInaccessible"
  | "TravelStatusNotAvailable"
  | "Unknown"

type ENPermissionStatus =
  | "Unknown"
  | "Active"
  | "Disabled"
  | "BluetoothOff"
  | "LocationOff"
  | "Restricted"
  | "Paused"
  | "Unauthorized"
```

mattThousand replied:
@devinjameson I agree the overlap is confusing; there are 2 separate concepts at play; one is the "status" of the `Exposure Notifications` framework, for which `BluetoothOff` is a valid state (neither a "success" nor a failure persay) and the other concept is the error that we might get back when attempting to activate ENs. 

For example, the `EN` framework could be in the `BluetoothOff` state, in which case if we try to activate it, we will get a `BluetoothOff` error response.

johnschoeman replied:
perhaps it makes sense to update the api for `requestAuthorization` to fulfill with only a 'success' or reject with the error?

mattThousand replied:
good points, I agree that the overlap is confusing. I think the intention in the current construction is to model the OS api; on the `iOS` side our JS method ultimately calls through to `setExposureNotificationEnabled` (https://developer.apple.com/documentation/exposurenotification/enmanager/3583729-setexposurenotificationenabled), which returns a `success` even in the case when `exposureNotificationStatus` is `bluetoothOff`; we _could_ move this special casing of the `bluetoothOff` scenario to the native layer, and vend a failure up to the JS layer in that case. Thoughts?

mattThousand replied:
https://github.com/Path-Check/gaen-mobile/pull/732

devinjameson committed:
WIP

devinjameson committed:
WIP

johnschoeman reviewed (approved):
Code looks good to me.

mattThousand reviewed (dismissed):
LGTM, let’s hold off on un-drafting until the android side is pushed

juancruzgs committed:
Return bluetooth and location status when EN status is requested (#730)

* Adjust iOS scoring threshold. (#727)

#### Why:
We'd like the scoring threshold check to reflect the original intention and be consistent across platforms.

#### This commit:
This commit brings the scoring threshold check for devices running `iOS 13.7+` in line with the Android implementation; the android implementation appropriately converts the the `weightedDurationSum` to minutes before making the comparison (https://github.com/Path-Check/gaen-mobile/blob/79e4d0091417fb71533edd3195d049b243be1fcd/android/app/src/main/java/org/pathcheck/covidsafepaths/exposurenotifications/ExposureNotificationClientWrapper.java#L229), and this commit performs the identical check in reverse, converting the `triggerThresholdWeightedDuration` value to seconds to perform the check

* Upgrade WorkManager to fix a crash (#726)

* Return bluetooth and location status when EN status is requested

Co-authored-by: Matt Buckley <matt.buckley@rocketinsights.com>

juancruzgs committed:
Return bluetooth and location status when EN status is requested (#730)

* Adjust iOS scoring threshold. (#727)

#### Why:
We'd like the scoring threshold check to reflect the original intention and be consistent across platforms.

#### This commit:
This commit brings the scoring threshold check for devices running `iOS 13.7+` in line with the Android implementation; the android implementation appropriately converts the the `weightedDurationSum` to minutes before making the comparison (https://github.com/Path-Check/gaen-mobile/blob/79e4d0091417fb71533edd3195d049b243be1fcd/android/app/src/main/java/org/pathcheck/covidsafepaths/exposurenotifications/ExposureNotificationClientWrapper.java#L229), and this commit performs the identical check in reverse, converting the `triggerThresholdWeightedDuration` value to seconds to perform the check

* Upgrade WorkManager to fix a crash (#726)

* Return bluetooth and location status when EN status is requested

Co-authored-by: Matt Buckley <matt.buckley@rocketinsights.com>

juancruzgs committed:
Convert EN errors to strings and pass them to JS (#731)

juancruzgs committed:
Convert EN errors to strings and pass them to JS (#731)

Commit:
Treat `BluetoothOff` as Error when requesting to enable Exposure Notifications (iOS) #732


Why:
When requesting to enable exposure notifications from the JS layer, we'd like to treat responses other than an Active status as errors.

This commit:
This commit updates the iOS exposure enablement call to vend a success to the JS layer only if the exposureNotificationStatus is Active after the request, and vends and error otherwise

Commit:
Treat `BluetoothOff` as Error when requesting to enable Exposure Notifications (iOS) #732


Why:
When requesting to enable exposure notifications from the JS layer, we'd like to treat responses other than an Active status as errors.

This commit:
This commit updates the iOS exposure enablement call to vend a success to the JS layer only if the exposureNotificationStatus is Active after the request, and vends and error otherwise

Commit:
work in progress

Commit:
work in progress

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

Commit:
Work in progress.

Commit:
Work in progress.

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson commented on src/Device/PermissionsContext.tsx:
```diff
@@ -46,12 +43,14 @@ export type ENPermissionStatus =
   | "Active"
   | "Disabled"
   | "BluetoothOff"
+  | "LocationOffAndRequired"
```
@mattThousand, changed the name of this because it clarifies that `LocationOffAndNotRequired` is not a possible state for `ENPermissionStatus`. We may want to change this on the native layer in some places, too?

devinjameson replied:
@juancruzgs This also may be relevant to the work you are doing tomorrow ^

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
Use doesDeviceSupportLocationlessScanning

devinjameson committed:
Use doesDeviceSupportLocationlessScanning

juancruzgs committed:
Check if locationless scanning is supported

juancruzgs committed:
Check if locationless scanning is supported

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

devinjameson committed:
WIP

Commit:
work in progress.

Commit:
work in progress.

Commit:
Merge branch 'dj-fix-location' of github.com:Path-Check/gaen-mobile into dj-fix-location

Commit:
work in progress.
